File: medium/smallest_trimmed_numbers.go
package medium

import (
	"slices"
)

type Rank struct {
	Index int
	Value string
}

func queryNumber(data []string, query []int) []Rank {
	result := make([]Rank, len(data))
	for i, word := range data {
		start := len(word) - query[1]
		if start < 0 {
			start = 0
		}
		result[i] = Rank{
			Index: i,
			Value: word[start:],
		}
	}
	return result
}
func SmallestTrimmedNumbers(nums []string, queries [][]int) []int {
	result := make([]int, len(queries))
	for i, query := range queries {
		trimmed := queryNumber(nums, query)
		slices.SortFunc(trimmed, func(a Rank, b Rank) int {
			if a.Value != b.Value {
				if a.Value < b.Value {
					return -1
				}
				if a.Value > b.Value {
					return 1
				}
				return 0
			}
			if a.Index != b.Index {
				return a.Index - b.Index
			}
			return 0
		})
		result[i] = trimmed[query[0]-1].Index
	}
	return result
}

File: lib/medium/smallest_trimmed_numbers.dart
class Rank {
  int index;
  String value;
  Rank(this.index, this.value);
  int getIndex() {
    return index;
  }

  String getValue() {
    return value;
  }
}

List<int> smallestTrimmedNumbers(List<String> nums, List<List<int>> queries) {
  List<int> result = List.filled(queries.length, 0);
  for (int i = 0; i < queries.length; i++) {
    int rank = queries[i][0];
    int digit = queries[i][1];
    List<Rank> trimmed =
        nums.indexed.map((e) {
          final idx = e.$1;
          final val = e.$2;
          final start = (val.length - digit).clamp(0, val.length);
          return Rank(idx, val.substring(start));
        }).toList();
    trimmed.sort((Rank lhs, Rank rhs) {
      int cmp = lhs.value.compareTo(rhs.value);
      if (cmp != 0) return cmp;
      return lhs.index.compareTo(rhs.index);
    });
    result[i] = trimmed[rank - 1].index;
  }
  return result;
}
